### ISSUE TITLE
Suggest a solution to confusing precedence issues related to Borrow / RefCell

### ISSUE THREAD

**Issue by hatf0** [rust-lang/rust-clippy#11040]
### What it does

This lint would catch a common mistake with `RefCell` / `Borrow`, where Borrow::borrow can confusingly take precedence over `RefCell::borrow`, resulting in incredibly unhelpful diagnostics that are unrelated to the problem (see: https://github.com/rust-lang/rust/issues/94858). 

Ideally this should be fixed in the compiler itself, but it would also be nice to have a lint to catch this issue.

### Example

```rust
use std::{borrow::BorrowMut, cell::RefCell, rc::Rc};

pub struct Foo {
    bar: (),
}

fn foo() {
    let foo = Rc::new(RefCell::new(Foo { bar: () }));
    // lint here: consider explicitly specifying either RefCell::borrow_mut or BorrowMut::borrow_mut
    let bar = foo.borrow_mut().bar;
}
```

Could be written as:

```rust
use std::{borrow::BorrowMut, cell::RefCell, rc::Rc};

pub struct Foo {
    bar: (),
}

fn foo() {
    let foo = Rc::new(RefCell::new(Foo { bar: () }));
    let bar = RefCell::borrow_mut(&foo).bar;
}
```

**Comment by y21** [comment]
This is something that I've run into myself a few times so a better diagnostic would be very nice, but I'm not sure if having a clippy lint would make sense.
This basically almost always results in typecheck errors (like in your example), and I don't think clippy gets to do anything if code fails to typecheck, so it would never actually get to run on code that does this

**Comment by Centri3** [comment]
I like the idea, but using this for a single method is a bit silly. Many types like `Rc`, `Arc`, etc. are supposed to be called like `Rc::clone` rather than `rc.clone()`. We should lint these cases

We can probably implement this as an early lint

**Comment by hatf0** [comment]
> I like the idea, but using this for a single method is a bit silly. Many types like `Rc`, `Arc`, etc. are supposed to be called like `Rc::clone` rather than `rc.clone()`. We should lint these cases

I agree! This issue definitely does exists with more than just `Borrow` & `RefCell`, but for brevity sake I only mentioned those two. The example in my original post shows that it also happens with `BorrowMut`, and I can imagine that it happens with other types. I guess a better title would be to lint in cases where there might be confusing precedence behavior?

**Comment by Alexendoo** [comment]
> I like the idea, but using this for a single method is a bit silly. Many types like Rc, Arc, etc. are supposed to be called like Rc::clone rather than rc.clone(). We should lint these cases

For `Rc::clone` it's identical either way, and just down to personal preference. I would say it's unrelated to the issue of a trait method surprisingly coming into scope

**Comment by Centri3** [comment]
Oh yeah that's a good point, `clone` was a bad example. Just the first thing that came to mind. Stuff like `downgrade` then ^^
